### PR TITLE
clang-tidy, clang-format and Jenkinsfile improvements

### PR DESCRIPTION
**Pull request by wiktorn**
1. Add target lint to CMake to check against Google checks
2. Add target format to CMake to check against errors in code formatting
3. Add lint and format targets to Jenkinsfile and WarningsPublisher
4. Improve performance of builds on Windows
5. Bring back pipes and tee when gathering errors and warnings for WarningsPublisher
6. Do not use XCode generator for CMake. That allows building on MacOS with only Command Line Developer Tools installed

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -464,6 +464,108 @@ endif(GTEST_FOUND AND GMOCK_FOUND)
...
+  # to COMMAND, for example by using "-I$<JOIN:${LINT_INC_DIRS}, -I>"
+  string(REPLACE ";" ";-I" LINT_INCLUDES "${LINT_INC_DIRS}")
+
+  # checks
+  list(APPEND LINT_CHECKS "google-*")
+
+  string(REPLACE ";" "," LINT_CHECKS_STR "${LINT_CHECKS}")
+  message(STATUS "clang-tidy includes: -I${LINT_INCLUDES} -I${CMAKE_BINARY_DIR} -I${PROJECT_SOURCE_DIR}")
+  message(STATUS "clang-tidy checks: ${LINT_CHECKS_STR}")
+  # parse CMAKE_CXX_FLAGS to list, so spaces will not be escaped in COMMAND's
```
Comment (mkow) [inline CMakeLists.txt]: `COMMAND's`? you sure you wanted `'s` here? ;)

File: CMakeLists.txt [hunk 2]
```diff
@@ -464,6 +464,108 @@ endif(GTEST_FOUND AND GMOCK_FOUND)
...
+  separate_arguments(LINT_CXX_FLAGS UNIX_COMMAND ${CMAKE_CXX_FLAGS})
+
+  # as a alternative - we could use
```
Comment (mkow) [inline CMakeLists.txt]: `an`

File: CMakeLists.txt [hunk 3]
```diff
@@ -464,6 +464,108 @@ endif(GTEST_FOUND AND GMOCK_FOUND)
...
+    message(STATUS "Looking for clang-tidy - not found")
+    message(STATUS "  Build target 'lint' will not be available.")
+endif()
+
+
+message(STATUS "Looking for clang-format")
+find_program(CLANG_FORMAT NAMES clang-format clang-format-4.0 clang-format-3.9 clang-format-3.8 clang-format-3.7 clang-format-3.6 clang-format-3.5)
+
+if(CLANG_FORMAT)
+  message(STATUS "Looking for clang-format- found")
```
Comment (mkow) [inline CMakeLists.txt]: space before `-`

File: CMakeLists.txt [hunk 4]
```diff
@@ -464,6 +464,108 @@ endif(GTEST_FOUND AND GMOCK_FOUND)
...
+    format
+    COMMAND
+      ${CLANG_FORMAT}
+      -style=Google
+      -sort-includes
+      -i
+      ${LINT_ALL_SOURCE_FILES}
+  )
+else()
+    message(STATUS "Looking for clang-format- not found")
```
Comment (mkow) [inline CMakeLists.txt]: ditto

File: Jenkinsfile [hunk 5]
```diff
@@ -1,4 +1,27 @@
 #!/usr/bin/env groovy
+// TODO: To get a more sensible workspace directory name
```
Comment (mkow) [inline Jenkinsfile]: `To get` -> `Get`

File: Jenkinsfile [hunk 6]
```diff
@@ -1,4 +1,27 @@
...
+/**
+ * This Jenkins file uses custom Warning parsers for clang-tidy output.
+ * Jenkins / Manage Jenkins / Configure System
+ * In section Compiler Warnings three parser needs to be configured:
```
Comment (mkow) [inline Jenkinsfile]: `Compiler Warnings` -> `"Compiler Warnings"`
`parser` -> `parsers`

File: Jenkinsfile [hunk 7]
```diff
@@ -1,4 +1,27 @@
...
+String message = matcher.group(4)
+Priority priority = Priority.NORMAL
+
+if (category.contains('error')) priority = Priority.HIGH
+
+return new Warning(fileName, lineNumber, "Clang-tidy", category, message, priority)
+
+ */
+
 // To get a more sensible workspace directory name
```
Comment (mkow) [inline Jenkinsfile]: Duplicated with "TODO" at the beginning?

File: Jenkinsfile [hunk 8]
```diff
@@ -47,48 +70,89 @@ builders['msvc2015_64'] = { node('windows'){
...
-              bat(script: "rd /s /q build_${version}", returnStatus: true)
-              bat script: "md build_${version}", returnStatus: true
-              dir ("build_${version}") {
-                bat script: "cmake -DCMAKE_PREFIX_PATH=%CMAKE_PREFIX_PATH% -DGOOGLETEST_SRC_PATH=%GOOGLETEST_DIR% -DEMBED_PYTHON_ARCHIVE_PATH=%EMBED_PYTHON_ARCHIVE_PATH% -DVCREDIST_BINARY=\"${vcredist_binary}\" -DOPENSSL_DLL_DIR=${tool 'openssl-64'} -G \"${generator}\" ..\\"
-                bat script: "cmake --build . --config ${buildConfiguration} > error_and_warnings.txt"
-                bat script: "type error_and_warnings.txt"
-                bat script: "cpack -D CPACK_PACKAGE_FILE_NAME=veles-${version} -G \"${cpack_generator}\" -C ${buildConfiguration}"
+            }
+          },
+          "python tests - Windows 10": {
```
Comment (mkow) [inline Jenkinsfile]: Is it Win10-only? Why?

File: Jenkinsfile [hunk 9]
```diff
@@ -142,114 +205,192 @@ builders['msvc2015'] = {node('windows'){
...
+              dir ("build") {
+                def branch = getBranch()
+                sh """cmake -DGOOGLETEST_SRC_PATH=\"${tool 'googletest'}\" -DCMAKE_BUILD_TYPE=${buildConfiguration} .."""
+                sh """#!/bin/bash -ex
+                  set -o pipefail
+                  cmake --build . --config ${buildConfiguration} -- -j3 2>&1 | tee error_and_warnings.txt
+                """
+                sh """#!/bin/bash -ex
+                  set -o pipefail
+                  cmake --build . --target lint --config ${buildConfiguration} -- -j3 2>&1 | tee lint_error_and_warnings.txt
```
Comment (mkow) [inline Jenkinsfile]: `lint_error_and_warnings` -> `lint_errors_and_warnings` (please fix this also in other places)

File: Jenkinsfile [hunk 10]
```diff
@@ -142,114 +205,192 @@ builders['msvc2015'] = {node('windows'){
...
+            unHealthy: ''
+          ])
         }
       } catch (error) {
-        post_stage_failure(env.JOB_NAME, "python-macosx", env.BUILD_ID, env.BUILD_URL)
+        post_stage_failure(env.JOB_NAME, "python-ubuntu1604", env.BUILD_ID,
+         env.BUILD_URL)
         throw error
-      }
+       }
```
Comment (mkow) [inline Jenkinsfile]: probably unintended

File: Jenkinsfile (hunk 8 above)
Comment (wiktorn) [inline Jenkinsfile]: It is where we run our tests. Actually, it is not Windows 10, but Windows 2016, so I just removed version reference.

**Comment by wiktorn** [comment]
Closes #259

**Pull request merged**